berth_reservations/relay.py

```diff
@@ -0,0 +1,22 @@
+from django.utils.translation import ugettext_lazy as _
+from graphene import relay
+from graphql_jwt.decorators import login_required, superuser_required
+
+from berth_reservations.exceptions import VenepaikkaGraphQLError
+
+
+class AuthorizationNode(relay.Node):
+    @classmethod
+    @login_required
+    @superuser_required
+    def get_node_from_global_id(cls, info, global_id, only_type=None):
+        node = super().get_node_from_global_id(info, global_id, only_type)
+
+        user = info.context.user
+        # TODO: implement proper permissions
+        if (hasattr(node, "user") and node.user == user) or user.is_superuser:
```

Comment:
hm, quite often it will be either `node.customer.user` or `node.owner.user` 🤔 a what a can a we a do? 🤔 🤔 🤔

Reply:
I think `AuthorizationNode` should somehow call child node's method, which would have proper check on the user... something along that line.

Reply:
Can a we a do a something a like:

```python
node_user = (hasattr(node, "user") and node.user) or \
    (hasattr(node, "customer") and node.customer.user) or \
    (hasattr(node, "owner") and node.owner.user)
```
???

Reply:
this seems like a hack IMO, since we can't cover all the possible cases here. we should just allow for sucessful auth checks in child nodes somehow and leave it to them to know what attribute has user connected to it

Reply:
It does feel a bit hacky. How about this: We add an optional method to the classes implementing `AuthorizationNode`, say 
```python
def authenticate_user(node):
```

and in `get_node_from_global_id` we can do something like:
```python
def get_node_from_global_id(cls, info, global_id, only_type=None)
    node = super().get_node_from_global_id(info, global_id, only_type)
    user = info.context.user

    is_own_node = False
    if hasattr(cls, "authenticate_user"):
        is_own_node = cls.authenticate_user(node)
    
    if is_own_node or user.is_superuser:
        return node
    else:
        raise VenepaikkaGraphQLError(
            _("You do not have permission to perform this action.")
        )
```

and leave the implementation details for handling the user to the implementing class.

Reply:
looks pretty good! cool idea 👍 
I am now trying to imagine if it can at all be done through the combination of `get_node_from_global_id` and `get_node`, without custom methods. 🤔

Reply:
Sorry to break your heart, but after some tinkering, it wouldn't be as straightforward as that.

One option would be to have pre-defined Authorization classes, depending on the user field:
```python
class UserAuthorizationNode(relay.Node):
    pass

class CustomerAuthorizationNode(relay.Node):
    pass

class OwnerAuthorizationNode(relay.Node):
    pass
```
but this might not be as scalable/maintainable.

---

Another option would be to have a custom class for every Node which is using `AuthorizationNode`, which would give us more granularity to check for specific things based on every resource, but we'd need to write and maintain even more custom classes.

```python
class BerthLeaseAuthorizationNode(AuthorizationNode):
    @classmethod
    def is_authorized(cls, node, user):
        return hasattr(node, "user") and node.user == user


class BerthLeaseNode(DjangoObjectType):
    berth = graphene.Field(BerthNode)
    status = LeaseStatusEnum()

    class Meta:
        model = BerthLease
        interfaces = (BerthLeaseAuthorizationNode,)
```

Reply:
I think we hit a dead end 🙈

Reply:
Discussed IRL, approach not needed after all, since just overriding `get_node` on each class that implements `graphene.relay.Node` and adding permission checks there is actually enough.

Reply:
That, and assuming that, in the case and invalid id is passed to the query, it will return
```json
{
  "data": {
    "query": null
  }
}
```

whether the user is logged in, has permissions, or not.

Reply:
slight correction: 

> invalid id is passed

means entirely undecodable Global ID, i.e. something like `id: "foobar"`.

when a user passes something that could be decoded (e.g. `QmVydGhBcHBsaWNhdGlvbk5vZGU6aW52YWxpZC1pZA==` can be decoded to `BerthApplicationNode:invalid-id`), then `get_node` does get called on the proper node class and the permission check fire first, giving the response ~"You do not have permissions"